### PR TITLE
minizip-ng: Add 3.0.0

### LINKED ISSUE #4386: [request] minizip-ng/3.0.0
### Package Details
  * Package Name/Version: **minizip-ng/3.0.0**
  * Website: **https://github.com/zlib-ng/minizip-ng**
  * Source code: **https://github.com/zlib-ng/minizip-ng**

### Description Of The Library / Tool
Fork of minizip (cross-platform zip manipulation library written in C) which already has recipe here with bugfixes and new features.

It has 738 stars currently.

### Additional information
Optional dependencies:
- bzip2
- iconv
- libbsd
- liblzma
- openssl
- zlib
- zstd

Probably some changes should be made in CMakeLists.txt for easier packaging and proper injection of dependencies

### PR DESCRIPTION
**Pull request by intelligide**
Specify library name and version:  **minizip-ng/3.0.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Closes #4386.

### CODE REVIEW

File: recipes/minizip-ng/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+from conans import ConanFile, tools, CMake
+
+
+class MinizipNgConan(ConanFile):
+    name = "minizip-ng"
+    description = "Fork of the popular zip manipulation library found in the zlib distribution."
+    topics = ("conan", "compression", "zip")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/zlib-ng/minizip-ng"
+    license = "BSD-3-Clause"
```
Comment (Nekto89) [inline recipes/minizip-ng/all/conanfile.py:11]:
```suggestion
    license = "Zlib"
```

File: recipes/minizip-ng/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        self.build_requires('pkgconf/1.7.3')
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
+
+        if tools.is_apple_os(self.settings.os) and self.options.with_libcomp:
+            self.options.with_zlib = False
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:68]:
```suggestion
        if tools.is_apple_os(self.settings.os) and self.options.with_libcomp:
            del self.options.with_zlib
```

If there is no option of zlib, why generate a package ID?

File: recipes/minizip-ng/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    def requirements(self):
+        if self.options.with_zlib:
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:71]:
```suggestion
        if self.options.get_safe("with_zlib"):
```

File: recipes/minizip-ng/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        extracted_dir = "minizip-ng-{}".format(self.version)
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["MZ_FETCH_LIBS"] = False
+        self._cmake.definitions["MZ_COMPAT"] = self.options.mz_compatibility
+        self._cmake.definitions["MZ_ZLIB"] = self.options.with_zlib
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:96]:
```suggestion
        self._cmake.definitions["MZ_ZLIB"] = self.options.get_safe("with_zlib", False)
```

File: recipes/minizip-ng/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        self._cmake.definitions["MZ_BZIP2"] = self.options.with_bzip2
+        self._cmake.definitions["MZ_LZMA"] = self.options.with_lzma
+        self._cmake.definitions["MZ_ZSTD"] = self.options.with_zstd
+        self._cmake.definitions["MZ_OPENSSL"] = self.options.with_openssl
+        self._cmake.definitions["MZ_LIBCOMP"] = tools.is_apple_os(self.settings.os) and self.options.with_libcomp
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:101]:
```suggestion
        self._cmake.definitions["MZ_LIBCOMP"] = tools.is_apple_os(self.settings.os) and self.options.with_libcomp
```
Does make sense keep `with_libcomp` for all configurations, if it is only used for iOS?

File: recipes/minizip-ng/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+class MinizipNgConan(ConanFile):
+    name = "minizip-ng"
+    description = "Fork of the popular zip manipulation library found in the zlib distribution."
+    topics = ("conan", "compression", "zip")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/zlib-ng/minizip-ng"
+    license = "Zlib"
+    exports_sources = "CMakeLists.txt"
+    generators = "cmake", "cmake_find_package", "pkg_config"
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:13]:
```suggestion
    generators = "cmake", "cmake_find_package", "pkg_config"
    provides = "minizip"
```

I think both `minizip` and `minizip-ng` can not be used together. We need to update `minizip` too.

Comment (prince-chrismc) [reply recipes/minizip-ng/all/conanfile.py:13]: Thats only true if the compatability option is True https://github.com/zlib-ng/minizip-ng/blob/6f17da582d2327c8cdc43622ddd31a185b6ef5fb/CMakeLists.txt#L604

File: recipes/minizip-ng/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
+
+        if not tools.is_apple_os(self.settings.os):
+            del self.options.with_libcomp
+        elif self.options.with_libcomp:
+            del self.options.with_zlib
```
Comment (prince-chrismc) [inline recipes/minizip-ng/all/conanfile.py:70]: This also needs to be in the validate I suspect https://github.com/zlib-ng/minizip-ng/blob/6f17da582d2327c8cdc43622ddd31a185b6ef5fb/CMakeLists.txt#L189

#4261

File: recipes/minizip-ng/all/conanfile.py (hunk 5 above)
Comment (prince-chrismc) [reply recipes/minizip-ng/all/conanfile.py:103]: It's only support on Apple https://github.com/zlib-ng/minizip-ng/blob/6f17da582d2327c8cdc43622ddd31a185b6ef5fb/CMakeLists.txt#L182

**Review by prince-chrismc — COMMENTED** [review]
Do these flags https://github.com/zlib-ng/minizip-ng/blob/6f17da582d2327c8cdc43622ddd31a185b6ef5fb/CMakeLists.txt#L647-L659 conflict with what conan will set?

File: recipes/minizip-ng/all/conanfile.py (hunk 6 above)
Comment (intelligide) [reply recipes/minizip-ng/all/conanfile.py:13]: `minizip` and `minizip-ng` produces both the same binaries (except on Windows) but doesn't provide the same API unless compatability option is true.

Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py:13]: So it can be added on configure method.

File: recipes/minizip-ng/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            del self.options.with_libbsd
+
+    def build_requirements(self):
+        self.build_requires('pkgconf/1.7.3')
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
```
Comment (uilianries) [inline recipes/minizip-ng/all/conanfile.py:66]:
```suggestion
        del self.settings.compiler.libcxx
        if self.options.mz_compatibility:
            self.provides = "minizip"
```

File: recipes/minizip-ng/all/conanfile.py (hunk 5 above)
Comment (uilianries) [reply recipes/minizip-ng/all/conanfile.py:104]: So it can be remove from package ID for other OSs, it's one item less to be computed on package ID.

**Comment by Nekto89** [comment]
> 
> 
> Do these flags https://github.com/zlib-ng/minizip-ng/blob/6f17da582d2327c8cdc43622ddd31a185b6ef5fb/CMakeLists.txt#L647-L659 conflict with what conan will set?

I had to remove them for local build through CMake's ExternalProject because it was overriding /Z7 option

**Pull request closed without merging**